---
title: "" 
author: "Kirsten Gotting"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    theme: cerulean
---



```{r}
## configuration

workingdir <-
control    <-
treatment  <- 
pval       <- 0.00001



setwd(workingdir)


source('scripts/runDEseq.R')
```


```{r initializeData, echo = FALSE, results = "hide", message = FALSE, error = FALSE, warning = FALSE}

## First list the R libraries I want to use
libraries        <- c('ggplot2', "DESeq2", "matrixStats", "gplots", "RColorBrewer", "parallel", "lattice", "plyr", "gtools", 'knitr', 'grid', 'gridExtra', "dplyr", 'tidyr', 'kiRsten')

## Now read in the libraries
lapply(libraries, function(x){
    library(x, character.only = TRUE, quietly = TRUE)
    })


## Initialize knitr options
opts_chunk$set(echo=FALSE, message=FALSE, results="hide", fig.keep="all", warning=FALSE, error=FALSE, fig.path="figures/")
opts_knit$set(root.dir = workingdir)


```




# Overview

Analyze the Mnemiopsis data with the new filtered transcriptome that Eric generated.



Working Directory:`r getwd()`


#Spearman Correlation



The Spearman correlation of the RPM (reads per million) for all genes that have an average RPM &#8805; 1. 

```{r spearman_correlation, fig.cap = ""}

## Make a spearman plot


filteredRPMs <- normCounts %>% group_by(ID) %>% mutate(rpmFlag = ifelse((mean(rpm) >= 1), yes = 'yes', no = 'no')) %>% ungroup() %>%
    filter(rpmFlag == 'yes') %>% select(ID, Sample, rpm) %>% spread(Sample, rpm)



rpmfilt        <- data.frame(select(filteredRPMs, -ID), row.names = filteredRPMs$ID, check.names = FALSE)
spearman       <- data.frame(cor(rpmfilt, method = 'spearman'), check.names = FALSE)
spearman$ID    <- rownames(spearman)
spearmanTidy   <- spearman %>% gather(Sample, correlation, -ID)

## and plot it!

ggplot(data = spearmanTidy, aes(x = ID, y = Sample, fill = correlation)) + geom_tile() +
    scale_fill_gradient2(low = "yellow", mid = 'white', high = "purple4", midpoint = 0.8, limit = c(0,1)) +
        ggtitle('Spearman Correlation of Samples') + xlab('') + ylab('') +
            theme(axis.text.x = element_text(angle = 90, hjust = 1))

## make a table of intra timepoint sample variations


intraSampleSpearman <- spearmanTidy %>% filter(ID != Sample) %>%
    separate(ID, into = c('Sample1Timepoint', 'Sample1Replicate'), sep = 4) %>% 
    separate(Sample, into = c('Sample2Timepoint', 'Sample2Replicate'), sep = 4) %>%
    filter(Sample1Timepoint == Sample2Timepoint) %>%
    group_by(Sample1Timepoint) %>%
    mutate(intraTimeCorr = mean(correlation)) %>%
    ungroup() %>%
    unite(Sample1, Sample1Timepoint, Sample1Replicate, sep = "") %>%
    unite(Sample2, Sample2Timepoint, Sample2Replicate, sep = "") 

```

Table of Spearman Correlation between samples, try sorting by the correlation to find the most/least correlated samples.

```{r, results='asis'}
DT::datatable(intraSampleSpearman)

```


#PCA plots

Principle Component Analysis on the log transformation of the count data.

## All Samples PCA
```{r PCA_plots,  fig.keep = "all", dev = "png"}

## make a PCA with all of the samples

library('broom')

pca_data   <- plotPCA(rld, returnData = TRUE, ntop = 10000)


pcaData <- function(rld, numKeep = 10000){
  pcaVar  <- data.frame(gene = rownames(assay(rld)), variance = rowVars(assay(rld)))
  top     <- arrange(pcaVar, desc(variance)) %>% slice(1:numKeep) %>% .$gene
  pcaDF   <- assay(rld)[top, ]
  prcomp(pcaDF)
}


pca_data <- pcaData(rld)


pca_rot <- data.frame(pca_data$rotation, Sample = rownames(pca_data$rotation), Condition = sub("_.*", "", rownames(pca_data$rotation)))

ggplot(data = pca_rot, aes(x = PC1, y = PC2, label = Sample, colour = factor(Condition))) +
  geom_point(size = 1) + 
  theme_bw() + 
  geom_text(hjust = 0, vjust = 0, size = 3) +
  xlab("PC 1") + 
  ylab("PC 2") + 
  ggtitle('All Samples PCA') + 
  theme(legend.position="none") + 
  scale_x_continuous(expand = c(0.2, 0)) 


ggplot(data = pca_rot, aes(x = PC2, y = PC3, label = Sample, colour = factor(Condition))) +
  geom_point(size = 1) + 
  theme_bw() + 
  geom_text(hjust = 0, vjust = 0, size = 3) +
  xlab("PC 2") + 
  ylab("PC 3") + 
  ggtitle('All Samples PCA') + 
  theme(legend.position="none") + 
  scale_x_continuous(expand = c(0.2, 0)) 


importance <- data.frame(Metric = rownames(summary(pca_data)$importance), summary(pca_data)$importance, row.names = NULL)
## print the individual PCA plots
```

```{r}

importanceDF <- gather(importance, Component, Value, -Metric) %>% filter(Metric == 'Proportion of Variance')

importanceDF$Component  <- factor(importanceDF$Component, levels = importanceDF$Component)

ggplot(importanceDF, aes(y = Value, x = Component)) + 
  geom_bar(stat = 'identity', fill = 'purple') + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ggtitle('Principle Component Proportion of Variance') + 
  ylab('Proportion of Variance')



```

Table of Principal Component Importance

```{r, results='asis'}

DT::datatable(importance[,1:11])

```


## Pairwise PCA

```{r}

individ.pcas <- lapply(contrast_names, function(x){
      pca_data   <- plotPCA(rld_all[[x]], returnData = TRUE, ntop = 10000)
      ggplot(data = pca_data, aes(x = PC1, y = PC2, label = name, colour = factor(condition))) +
        geom_point(size = 3) + 
        theme_bw() + 
        geom_text(hjust = 0, vjust = -0.4) +
        xlab("PC 1") +
        ylab("PC 2") + 
        ggtitle(x) + 
        scale_x_continuous(expand = c(0.3, 0))  +
        theme(legend.position="none") 
})

print(individ.pcas)

```


#Pairwise Analysis


## MA and Volcano plots

MA plot: The log2 fold change(M) plotted against the log2 average(A) of the normalized read count for each gene.  Significantly differentially expressed genes are plotted in red.

Volcano plot: The log2 fold change(M) plotted against the -log10 (eg. 1e^-10 = 10) of the adjusted p-value.


```{r pairwisePlots, fig.keep = "all", dev = "png", fig.height = 10, fig.width = 20}

##MA plots

##The log2 fold change(M) plotted against the log2 average(A) of the normalized read count for each gene.  Significantly differentially expressed genes are plotted in red.

##x <- contrast_names[1]

## make MA and Volcano plots


plots <- lapply(contrast_names, function(x){
    df      <- all_results_tidyDF %>% filter(contrastID == x) %>% spread(dea_ID, dea_Value) %>% mutate(logbaseMean = log(baseMean))
    title   <- x
    pval    <- as.numeric(pval)
    if(is.null(df)){
        message(paste0("Pvalue too stringent for ", contrast, ". MA/Volcano plot ommited.\n"))
    } else {
        pval                  <- as.numeric(pval)
        lower_label           <- paste0("Genes with Counts: ", nrow(df), "\nUp: ",#create the label of the legend
                                  nrow(subset(df, log2FoldChange > 0 & padj < pval)),
                                  ", Down: ", nrow(subset(df, log2FoldChange < 0 & padj < pval)), "\n")
  
        df <- df %>% mutate(significance_group = ifelse(padj <= pval, yes = 'yes', no = 'no'))
  
        legend_labels        <- c(paste0("padj > ", pval,": ", nrow(filter(df, significance_group == 'no'))),
                                  paste0("padj < ", pval,": ", nrow(filter(df, significance_group == 'yes'))))
        ma.plot <-   ggplot(df, aes(x = logbaseMean, y = log2FoldChange, group = significance_group, colour = significance_group)) +
                      geom_point() + theme_bw() +
                      ggtitle('MA plot') + scale_x_continuous(expand = c(0, 0)) +
                      xlab("A = Mean of log2(Abundance)") + ylab("M = log2(Fold Change)") +
                      scale_color_manual(values = c("gray24", "red"), name = lower_label, breaks = c("gray24", "red"), labels = legend_labels) +
                      geom_hline(aes(yintercept = 0), colour = "black", linetype = "dashed") +
                      geom_hline(aes(yintercept = 1), colour = "gray66", linetype = "dashed") +
                      geom_hline(aes(yintercept = -1), colour = "gray66", linetype="dashed") +
                      theme(legend.position = 'none', text = element_text(size=20))
        volcano.plot <-   ggplot(df, aes(x = log2FoldChange, y = -log(padj, base = c(10)), group = significance_group, colour = significance_group)) +
                          geom_point() +
                          ggtitle('Volcano plot') +
                          xlab("M = log2(Fold Change)") + ylab("-log10(Adjusted P-Value)") + 
                          scale_color_manual(values = c("gray24", "red"), labels = legend_labels, name=lower_label) +
                          theme_bw() +
                          theme(legend.title = element_text(size = 16, face = "bold"), legend.text = element_text(size = 20),  text = element_text(size=20))
          
      grid.arrange(arrangeGrob(ma.plot,
                              volcano.plot,
                              top = textGrob(title, gp = gpar(fonsize = 40, fontface = 'bold', cex = 2), hjust = 1.8),
                              nrow = 1, ncol = 2,
                              widths=c(0.38, 0.62)
      ))
    }
})

print(plots)


```


##Heatmaps



The first heatmap was made by hierarchically clustering the euclidean distances of the log2 fold change for genes with p-values less than `r pval` at any time point.

The second heatmap uses the same hierarchical method to cluster timepoints.


```{r pairwiseHeatmaps, results = "hide", fig.cap = "", fig.keep = "high"}


prettycolors  <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
col.breaks    <- seq(-3, 3, length.out = ncol(heatmap.input.table))


## Make two heatmaps, allow custimizable graphing

hr <- hclust(dist(heatmap.input.table), method="complete") # Creates row dendrogram
hc <- hclust(dist(t(heatmap.input.table)), method="complete") # Creates column dendrogram

myclr     <- cutree(hr, k = 7) # set up where the groups will be, change 'k' for more or less groups
mycolr    <- c("Green", "Red", "Blue", "Orange", "Purple", "Pink", "Yellow") # must be the same length as 'k'
mycolr    <- mycolr[as.vector(myclr)] # pull out the colors assigned to each gene
mycolr.df <- data.frame(Cluster_Color = mycolr, TranscriptID = names(myclr)) # create a data frame of the cluster color each gene belongs to


if(ncol(heatmap.input.table) > 1){
    ## Make a heatmap without a column dendrogram
    col.breaks    <- seq(-3, 3, length.out = ncol(heatmap.input.table))
    heatmap       <- heatmap.2(heatmap.input.table,
                               margin = c(15,5), 
                               col = prettycolors, 
                               dendrogram = 'row', 
                               trace = c("none"), 
                               Rowv=as.dendrogram(hr), 
                               Colv = NA, 
                               labRow = FALSE, 
                               key = TRUE, 
                               breaks = col.breaks, 
                               key.ylab = "Gene Count", 
                               key.xlab = "log2FoldChange", 
                               cexCol = 1.5, 
                               RowSideColors = mycolr)
    title(main = paste0("Clustering of ", nrow(heatmap.input.table)," Genes"), line = 0)
    par(cex.main=0.9)
}


if(ncol(heatmap.input.table) > 2) {
    k <- ifelse(ncol(heatmap.input.table < 5), 3, 5)
    myclc  <- cutree(hc, k = k) # create groupings for the column dendrogram
    mycolc <- c("Green", "Red", "Blue", "Orange", "Purple")
    mycolc <- mycolc[as.vector(myclc)]
    ## Make a heatmap of the logfold changes with column dendrogram.
    heatmap       <- heatmap.2(heatmap.input.table, 
                               margin = c(15,5), 
                               dendrogram = 'both', 
                               Rowv = as.dendrogram(hr), 
                               Colv = as.dendrogram(hc), 
                               col = prettycolors, 
                               trace = c("none"), 
                               labRow = FALSE, 
                               key = TRUE, 
                               breaks = col.breaks, 
                               main = paste0("Clustering of Timepoints on ", nrow(heatmap.input.table)," Genes"), 
                               key.ylab = "Gene Count", 
                               key.xlab = "log2FoldChange", 
                               cexCol = 1.5, 
                               RowSideColors = mycolr, 
                               ColSideColors = mycolc)
}

```



```{r save}
## save everything 
save.image(file= paste0("./workspace_images/full_analysis.RData"))

```



# GLM Analysis


## MA and Volcano plots

MA plot: The log2 fold change(M) plotted against the log2 average(A) of the normalized read count for each gene.  Significantly differentially expressed genes are plotted in red.

Volcano plot: The log2 fold change(M) plotted against the -log10 (eg. 1e^-10 = 10) of the adjusted p-value.


```{r glmPlots, fig.keep = "all", dev = "png", fig.height = 10, fig.width = 20}

##MA plots



##The log2 fold change(M) plotted against the log2 average(A) of the normalized read count for each gene.  Significantly differentially expressed genes are plotted in red.

## make MA and Volcano plots


plots <- lapply(contrast_names, function(x){
    df      <- all_results_glmTidy %>% filter(contrastID == x) %>% spread(dea_ID, dea_Value) %>% mutate(logbaseMean = log(baseMean))
    title   <- x
    pval    <- as.numeric(pval)
    if(is.null(df)){
        message(paste0("Pvalue too stringent for ", contrast, ". MA/Volcano plot ommited.\n"))
    } else {
        pval                  <- as.numeric(pval)
        lower_label           <- paste0("Genes with Counts: ", nrow(df), "\nUp: ",#create the label of the legend
                                  nrow(subset(df, log2FoldChange > 0 & padj < pval)),
                                  ", Down: ", nrow(subset(df, log2FoldChange < 0 & padj < pval)), "\n")
  
        df <- df %>% mutate(significance_group = ifelse(padj <= pval, yes = 'yes', no = 'no'))
  
        legend_labels        <- c(paste0("padj > ", pval,": ", nrow(filter(df, significance_group == 'no'))),
                                  paste0("padj < ", pval,": ", nrow(filter(df, significance_group == 'yes'))))
        ma.plot <-   ggplot(df, aes(x = logbaseMean, y = log2FoldChange, group = significance_group, colour = significance_group)) +
                      geom_point() + theme_bw() +
                      ggtitle('MA plot') +
                      xlab("A = Mean of log2(Abundance)") + ylab("M = log2(Fold Change)") +
                      scale_color_manual(values = c("gray24", "red"), name = lower_label, breaks = c("gray24", "red"), labels = legend_labels) +
                      geom_hline(aes(yintercept = 0), colour = "black", linetype = "dashed") +
                      geom_hline(aes(yintercept = 1), colour = "gray66", linetype = "dashed") +
                      geom_hline(aes(yintercept = -1), colour = "gray66", linetype="dashed") +
                      theme(legend.position = 'none', text = element_text(size=20))
        volcano.plot <-   ggplot(df, aes(x = log2FoldChange, y = -log(padj, base = c(10)), group = significance_group, colour = significance_group)) +
                          geom_point() +
                          ggtitle('Volcano plot') +
                          xlab("M = log2(Fold Change)") + ylab("-log10(Adjusted P-Value)") + 
                          scale_color_manual(values = c("gray24", "red"), labels = legend_labels, name=lower_label) +
                          theme_bw() +
                          theme(legend.title = element_text(size = 16, face = "bold"), legend.text = element_text(size = 20),  text = element_text(size=20))
          
      grid.arrange(arrangeGrob(ma.plot,
                              volcano.plot,
                              top = textGrob(title, gp = gpar(fonsize = 40, fontface = 'bold', cex = 2), hjust = 1.8),
                              nrow = 1, ncol = 2,
                              widths=c(0.38, 0.62)
      ))
    }
})

print(plots)


```




```{r glmHeatmaps, results = "hide", fig.cap = "", fig.keep = "high", eval = FALSE}
##Heatmaps


#The first heatmap was made by hierarchically clustering the euclidean distances of the log2 fold change for genes with p-values less than `r pval` at any time point.

#The second heatmap uses the same hierarchical method to cluster timepoints.

## Make two heatmaps

hr <- hclust(dist(heatmap.input.table), method="complete") # Creates row dendrogram
hc <- hclust(dist(t(heatmap.input.table)), method="complete") # Creates column dendrogram

myclr     <- cutree(hr, k = 7) # set up where the groups will be, change 'k' for more or less groups
mycolr    <- c("Green", "Red", "Blue", "Orange", "Purple", "Pink", "Yellow") # must be the same length as 'k'
mycolr    <- mycolr[as.vector(myclr)] # pull out the colors assigned to each gene
mycolr.df <- data.frame(Cluster_Color = mycolr, TranscriptID = names(myclr)) # create a data frame of the cluster color each gene belongs to



if(ncol(heatmap.input.table) > 1){
    ## Make a heatmap without a column dendrogram
    prettycolors  <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
    col.breaks    <- seq(-3, 3, length.out = ncol(heatmap.input.table))
    heatmap       <- heatmap.2(heatmap.input.table, margin = c(15,5), col = prettycolors, dendrogram = 'row', trace = c("none"), Rowv=as.dendrogram(hr), Colv = NA, labRow = FALSE, key = TRUE, breaks = col.breaks, key.ylab = "Gene Count", key.xlab = "log2FoldChange", cexCol = 1.5, RowSideColors = mycolr)
    title(main = paste0("Clustering of ", nrow(heatmap.input.table)," Genes"), line = 0)
    par(cex.main=0.9)
}


if(ncol(heatmap.input.table) > 2) {
    k <- ifelse(ncol(heatmap.input.table < 5), 3, 5)
    myclc  <- cutree(hc, k = k) # create groupings for the column dendrogram
    mycolc <- c("Green", "Red", "Blue", "Orange", "Purple")
    mycolc <- mycolc[as.vector(myclc)]
    ## Make a heatmap of the logfold changes with column dendrogram.
    prettycolors  <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
    col.breaks    <- seq(-3, 3, length.out = ncol(heatmap.input.table))
    heatmap       <- heatmap.2(heatmap.input.table, margin = c(15,5), dendrogram = 'both', Rowv = as.dendrogram(hr), Colv = as.dendrogram(hc), col = prettycolors, trace = c("none"), labRow = FALSE, key = TRUE, breaks = col.breaks, main = paste0("Clustering of Timepoints on ", nrow(heatmap.input.table)," Genes"), key.ylab = "Gene Count", key.xlab = "log2FoldChange", cexCol = 1.5, RowSideColors = mycolr, ColSideColors = mycolc)
}

```


```{r sessionInfo, echo=FALSE}
dir.create('sessionInfo')
writeLines(capture.output(sessionInfo()), "sessionInfo/QCsessionInfo.txt")
```

#[R-session information](`r paste0("/sessionInfo/QCsessionInfo.txt")`)

